community/io/src/main/java/org/neo4j/io/pagecache/impl/muninn/OptiLock.java

```diff
@@ -0,0 +1,229 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.io.pagecache.impl.muninn;
+
+import org.neo4j.unsafe.impl.internal.dragons.UnsafeUtil;
+
+/**
+ * OptiLock is a sequence-based lock like StampedLock, but with the ability to take optimistic write-locks.
+ * <p>
+ * The OptiLock supports non-blocking optimistic concurrent read locks, non-blocking concurrent write locks,
+ * and a pessimistic non-blocking exclusive lock.
+ * <p>
+ * The optimistic read lock works through validation, so at the end of the critical section, the read lock has to be
+ * validated and, if the validation fails, the critical section has to be retried. The read-lock acquires a stamp
+ * at the start of the critical section, which is then validated at the end of the critical section. The stamp is
+ * invalidated if any write-lock or exclusive lock has been taking since the stamp was acquired.
+ * <p>
+ * The optimistic write locks works by assuming that writes are always non-conflicting, so no validation is required.
+ * However, the write locks will check if a pessimistic exclusive lock is held at the start of the critical section,
+ * and if so, fail to be acquired. The write locks will invalidate all optimistic read locks. The write lock is
+ * try-lock only, and will never block.
+ * <p>
+ * The exclusive lock will also invalidate the optimistic read locks, but not the write locks. The exclusive lock is
+ * try-lock only, and will never block.
+ */
+public class OptiLock
```

Comment:
Class name is a bit cryptic and `keep an air of mystery about it`. Even if you can read a java doc and get the general idea why it's called like that would propose to rename it to OptimisticWriteStampedLock or similar